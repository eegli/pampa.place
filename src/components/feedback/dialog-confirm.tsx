import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
} from '@mui/material';

export type ConfirmationDialogProps = {
  title: string;
  infoMessage?: string;
  onCancelCallback: () => void;
  onConfirmCallback: () => void | (() => Promise<void>);
  confirmMessage?: string;
  cancelMessage?: string;
};

export const ConfirmationDialog = ({
  title,
  infoMessage,
  onCancelCallback,
  onConfirmCallback,
  confirmMessage = 'Confirm',
  cancelMessage = 'Cancel',
}: ConfirmationDialogProps) => {
  return (
    <Dialog open={true} onClose={onCancelCallback} PaperProps={{elevation: 2}}>
      <DialogTitle>{title}</DialogTitle>
      <DialogContent>
        {infoMessage && <DialogContentText>{infoMessage}</DialogContentText>}
      </DialogContent>
      <DialogActions>
        <Button onClick={onCancelCallback}>{cancelMessage}</Button>
        <Button onClick={onConfirmCallback} autoFocus>
          {confirmMessage}
        </Button>
      </DialogActions>
    </Dialog>
  );
};
