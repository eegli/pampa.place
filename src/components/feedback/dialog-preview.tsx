import {
  Box,
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  useMediaQuery,
  useTheme,
} from '@mui/material';
import {ReactNode} from 'react';

type PreviewDialogProps = {
  onCloseCallback: () => void;
  children: ReactNode;
  title: string;
  text?: string;
};

export const PreviewDialog = (props: PreviewDialogProps) => {
  const theme = useTheme();
  const fullScreen = useMediaQuery(theme.breakpoints.down('sm'));
  const {text, title, children, onCloseCallback} = props;
  return (
    <Dialog
      open={true}
      fullScreen={fullScreen}
      onClose={onCloseCallback}
      PaperProps={{elevation: 1}}
      sx={{
        borderRadius: 10,
      }}
    >
      <DialogTitle>{title}</DialogTitle>
      <DialogContent
        sx={{
          display: 'flex',
          flexDirection: 'column',
        }}
      >
        <Box
          id={`preview-dialog`}
          sx={{
            height: 600,
            maxHeight: '100%',
            width: 400,
            margin: 'auto',
            maxWidth: '90%',
          }}
        >
          {children}
        </Box>
        {text && <DialogContentText p={2}>{text}</DialogContentText>}
      </DialogContent>
      <DialogActions>
        <Button onClick={onCloseCallback}>Close</Button>
      </DialogActions>
    </Dialog>
  );
};
