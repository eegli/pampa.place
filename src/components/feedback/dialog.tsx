import {
  Button,
  Dialog as MuiDialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
} from '@mui/material';

export type DialogProps = {
  title: string;
  infoMessage: string;
  onCancelCallback: () => void;
  onConfirmCallback: () => void;
  confirmMessage?: string;
  cancelMessage?: string;
};

export const Dialog = ({
  title,
  infoMessage,
  onCancelCallback,
  onConfirmCallback,
  confirmMessage = 'Confirm',
  cancelMessage = 'Cancel',
}: DialogProps) => {
  return (
    <MuiDialog open PaperProps={{elevation: 1}}>
      <DialogTitle>{title}</DialogTitle>
      <DialogContent>
        <DialogContentText>{infoMessage}</DialogContentText>
      </DialogContent>
      <DialogActions sx={{mb: 2}}>
        <Button onClick={onCancelCallback} sx={{mr: 2}}>
          {cancelMessage}
        </Button>
        <Button
          variant="contained"
          onClick={onConfirmCallback}
          autoFocus
          sx={{mr: 2}}
        >
          {confirmMessage}
        </Button>
      </DialogActions>
    </MuiDialog>
  );
};
