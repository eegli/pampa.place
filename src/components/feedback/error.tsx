import {Alert, AlertColor, AlertTitle, Box, Button} from '@mui/material';
import Link from 'next/link';

type ErrorPros = {
  primaryAction?: () => void;
  primaryActionText?: string;
  secondaryAction?: () => void;
  secondaryActionText?: string;
  error: string;
  info?: string;
  title?: string;
  severity?: AlertColor;
};

export const Error = ({
  primaryAction,
  primaryActionText,
  secondaryAction,
  secondaryActionText,
  title = 'Something went wrong',
  info,
  error,
  severity = 'error',
}: ErrorPros) => {
  return (
    <Box padding={4} margin="auto">
      <Alert variant="outlined" severity={severity}>
        <AlertTitle>{title}</AlertTitle>
        {info && <p>{info}</p>}
        <p>
          <strong>Message:</strong>
        </p>
        <pre style={{whiteSpace: 'pre-wrap', overflow: 'auto'}}>{error}</pre>
      </Alert>

      <Box
        textAlign="center"
        display="flex"
        justifyContent="space-between"
        sx={{mt: 2}}
      >
        {secondaryAction ? (
          <Button onClick={secondaryAction}>{secondaryActionText}</Button>
        ) : (
          <Link href="/" passHref>
            <Button>Home</Button>
          </Link>
        )}

        {primaryAction && primaryActionText && (
          <Button variant="contained" onClick={primaryAction} sx={{mr: 3}}>
            {primaryActionText}
          </Button>
        )}
      </Box>
    </Box>
  );
};
