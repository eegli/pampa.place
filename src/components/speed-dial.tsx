import DarkModeIcon from '@mui/icons-material/DarkMode';
import FlagIcon from '@mui/icons-material/Flag';
import HomeIcon from '@mui/icons-material/Home';
import LightModeIcon from '@mui/icons-material/LightMode';
import RestartAltIcon from '@mui/icons-material/RestartAlt';
import ShortcutIcon from '@mui/icons-material/Shortcut';
import Box from '@mui/material/Box';
import SpeedDial from '@mui/material/SpeedDial';
import SpeedDialAction from '@mui/material/SpeedDialAction';
import { useRouter } from 'next/router';
import { useState } from 'react';
import { useAppDispatch, useAppSelector } from '../redux/hooks';
import { toggleTheme } from '../redux/slices/app';
import { RootState } from '../redux/store';
import Dialog, { DialogProps } from './dialog';

export default function SpeedDialTooltipOpen() {
  const [dialog, setDialog] = useState<DialogProps>({} as DialogProps);

  const router = useRouter();

  const homeDialog: DialogProps = {
    title: 'Abort the game and return home?',
    callbackTitle: 'Abort game',
    callback: function () {
      router.push('/');
    },
    open: true,
    cancel: function () {
      setDialog({ ...dialog, open: false });
    },
  };

  const restartDialog: DialogProps = {
    title: 'Restart the game?',
    callbackTitle: 'Restart',
    callback: function () {
      router.push('/');
    },
    open: true,
    cancel: function () {
      setDialog({ ...dialog, open: false });
    },
  };

  // TODO
  const giveUpDialog: DialogProps = {
    title: 'Give up and skip round?',
    callbackTitle: 'Give up',
    callback: function () {
      router.push('/');
    },
    open: true,
    cancel: function () {
      setDialog({ ...dialog, open: false });
    },
  };

  const activeTheme = useAppSelector((s: RootState) => s.app.theme);
  const dispatch = useAppDispatch();

  return (
    <>
      <Box
        id="speed-dial-menu-overlay"
        zIndex={100}
        position="fixed"
        bottom={30}
        left={30}
        sx={{
          height: 330,
          transform: 'translateZ(0px)',
          flexGrow: 1,
        }}
      >
        <SpeedDial
          ariaLabel="SpeedDial tooltip example"
          sx={{ position: 'absolute', bottom: 16, left: 16, zIndex: 100 }}
          icon={<ShortcutIcon />}
        >
          <SpeedDialAction
            key="Give up"
            icon={<FlagIcon />}
            tooltipTitle="Give up"
            tooltipPlacement="right"
            onClick={() => setDialog(giveUpDialog)}
          />
          <SpeedDialAction
            key="Restart"
            icon={<RestartAltIcon />}
            tooltipTitle="Restart"
            tooltipPlacement="right"
            onClick={() => setDialog(restartDialog)}
          />
          <SpeedDialAction
            key="Home"
            icon={<HomeIcon />}
            tooltipTitle="Home"
            tooltipPlacement="right"
            onClick={() => setDialog(homeDialog)}
          />
          <SpeedDialAction
            key="Mode"
            icon={
              activeTheme === 'light' ? <DarkModeIcon /> : <LightModeIcon />
            }
            tooltipTitle={activeTheme === 'light' ? 'Lights off' : 'Lights on'}
            tooltipPlacement="right"
            onClick={() => dispatch(toggleTheme())}
          />
        </SpeedDial>
      </Box>
      <Dialog {...dialog} />
    </>
  );
}
