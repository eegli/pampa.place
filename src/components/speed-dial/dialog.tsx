import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
} from '@mui/material';

export type DialogProps = {
  open: boolean;
  title: string;
  callback: () => void | (() => Promise<void>);
  callbackTitle: string;
  cancel: () => void;
  message?: string;
};

const ConfirmationDialog = ({
  open,
  title,
  callback,
  cancel,
  message,
  callbackTitle,
}: DialogProps) => {
  // Needed for smooth unmount transitions
  if (!open) {
    return <div />;
  }
  return (
    <div>
      <Dialog
        open={open}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
        PaperProps={{ elevation: 2 }}
      >
        <DialogTitle id="alert-dialog-title">{title}</DialogTitle>
        {message && (
          <DialogContent>
            <DialogContentText id="alert-dialog-description">
              {message}
            </DialogContentText>
          </DialogContent>
        )}
        <DialogActions>
          <Button onClick={cancel}>Cancel</Button>
          <Button onClick={callback} autoFocus>
            {callbackTitle}
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  );
};

export default ConfirmationDialog;
