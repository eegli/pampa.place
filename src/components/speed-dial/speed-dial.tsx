import {toggleTheme} from '@/redux/app/app.slice';
import {resetRound} from '@/redux/game/game.slice';
import {useAppDispatch, useAppSelector} from '@/redux/redux.hooks';
import {RootState} from '@/redux/redux.store';
import DarkModeIcon from '@mui/icons-material/DarkMode';
import FlagIcon from '@mui/icons-material/Flag';
import HomeIcon from '@mui/icons-material/Home';
import LightModeIcon from '@mui/icons-material/LightMode';
import RestartAltIcon from '@mui/icons-material/RestartAlt';
import ShortcutIcon from '@mui/icons-material/Shortcut';
import Box from '@mui/material/Box';
import SpeedDial from '@mui/material/SpeedDial';
import SpeedDialAction from '@mui/material/SpeedDialAction';
import {useRouter} from 'next/router';
import {useState} from 'react';
import Dialog, {DialogProps} from './dialog';

export default function SpeedDialTooltipOpen() {
  const [dialog, setDialog] = useState<DialogProps>({} as DialogProps);

  const router = useRouter();

  const activeTheme = useAppSelector((s: RootState) => s.app.theme);
  const dispatch = useAppDispatch();

  return (
    <>
      <Box
        id="speed-dial-menu-overlay"
        zIndex={100}
        position="fixed"
        bottom={30}
        left={30}
        sx={{
          height: 330,
          transform: 'translateZ(0px)',
          flexGrow: 1,
        }}
      >
        <SpeedDial
          ariaLabel="SpeedDial tooltip example"
          sx={{position: 'absolute', bottom: 16, left: 16, zIndex: 100}}
          icon={<ShortcutIcon />}
        >
          <SpeedDialAction
            key="Give up"
            icon={<FlagIcon />}
            tooltipTitle="Give up"
            tooltipPlacement="right"
            onClick={() =>
              setDialog({
                title: 'Give up and skip round?',
                callbackTitle: 'Give up',
                callback: function () {
                  router.push('/');
                },
                open: true,
                cancel: function () {
                  setDialog({...dialog, open: false});
                },
              })
            }
          />
          <SpeedDialAction
            key="Restart"
            icon={<RestartAltIcon />}
            tooltipTitle="Restart round"
            tooltipPlacement="right"
            onClick={() =>
              setDialog({
                title: 'Restart round?',
                message:
                  'This will reset the current round progress for all players. The first player will start the current round again in a new location.',
                callbackTitle: 'Restart round',
                callback: function () {
                  dispatch(resetRound());
                  setDialog({...dialog, open: false});
                },
                open: true,
                cancel: function () {
                  setDialog({...dialog, open: false});
                },
              })
            }
          />
          <SpeedDialAction
            key="Home"
            icon={<HomeIcon />}
            tooltipTitle="Home"
            tooltipPlacement="right"
            onClick={() =>
              setDialog({
                title: 'Abort the game and return home?',
                callbackTitle: 'Abort game',
                callback: function () {
                  router.push('/');
                },
                open: true,
                cancel: function () {
                  setDialog({...dialog, open: false});
                },
              })
            }
          />
          <SpeedDialAction
            key="Mode"
            icon={
              activeTheme === 'light' ? <DarkModeIcon /> : <LightModeIcon />
            }
            tooltipTitle={activeTheme === 'light' ? 'Lights off' : 'Lights on'}
            tooltipPlacement="right"
            onClick={() => dispatch(toggleTheme())}
          />
        </SpeedDial>
      </Box>
      <Dialog {...dialog} />
    </>
  );
}
