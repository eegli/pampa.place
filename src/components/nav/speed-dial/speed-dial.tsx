import {Dialog, DialogProps} from '@/components/feedback/dialog';
import {setTheme} from '@/redux/app';
import {resetRound} from '@/redux/game';
import {useAppDispatch, useAppSelector} from '@/redux/hooks';
import DarkModeIcon from '@mui/icons-material/DarkMode';
import HomeIcon from '@mui/icons-material/Home';
import LightModeIcon from '@mui/icons-material/LightMode';
import RestartIcon from '@mui/icons-material/RestartAlt';
import ShortcutIcon from '@mui/icons-material/Shortcut';
import Box from '@mui/material/Box';
import SpeedDial from '@mui/material/SpeedDial';
import SpeedDialAction from '@mui/material/SpeedDialAction';
import {useRouter} from 'next/router';
import {useState} from 'react';

type DialogStateProps = Omit<DialogProps, 'onCancelCallback'>;

export const SpeedDialNav = () => {
  const dispatch = useAppDispatch();
  const router = useRouter();
  const activeTheme = useAppSelector(s => s.app.theme);
  const [dialog, setDialog] = useState<DialogStateProps | undefined>(undefined);

  const dialogProps: DialogProps | undefined = dialog && {
    ...dialog,
    onCancelCallback: () => setDialog(undefined),
  };

  function handleToggleTheme() {
    dispatch(setTheme(activeTheme === 'light' ? 'dark' : 'light'));
  }

  return (
    <>
      <Box
        id="speed-dial-menu-overlay"
        zIndex={100}
        position="fixed"
        bottom={30}
        left={30}
        sx={{
          height: 330,
          transform: 'translateZ(0px)',
          flexGrow: 1,
        }}
      >
        <SpeedDial
          ariaLabel="speed-dial-menu"
          sx={{position: 'absolute', bottom: 16, left: 16, zIndex: 100}}
          icon={<ShortcutIcon />}
        >
          <SpeedDialAction
            key="restart"
            aria-label="restart"
            icon={<RestartIcon />}
            tooltipTitle="Restart round"
            tooltipPlacement="right"
            onClick={() => {
              setDialog({
                title: 'Restart round?',
                infoMessage:
                  'This will reset the current round progress for all players. The first player will start the current round again in a new location.',
                confirmMessage: 'Restart round',
                onConfirmCallback: function () {
                  setDialog(undefined);
                  dispatch(resetRound());
                },
              });
            }}
          />
          <SpeedDialAction
            key="home"
            aria-label="home"
            icon={<HomeIcon />}
            tooltipTitle="Home"
            tooltipPlacement="right"
            onClick={() => {
              setDialog({
                title: 'Abort the game and return home?',
                infoMessage: 'This will reset the current game',
                confirmMessage: 'Abort game',
                onConfirmCallback: async function () {
                  setDialog(undefined);
                  await router.push('/');
                },
              });
            }}
          />
          <SpeedDialAction
            key="theme"
            aria-label="toggle-theme"
            icon={
              activeTheme === 'light' ? <DarkModeIcon /> : <LightModeIcon />
            }
            tooltipTitle={activeTheme === 'light' ? 'Lights off' : 'Lights on'}
            tooltipPlacement="right"
            onClick={handleToggleTheme}
          />
        </SpeedDial>
      </Box>

      {dialogProps && <Dialog {...dialogProps} />}
    </>
  );
};
