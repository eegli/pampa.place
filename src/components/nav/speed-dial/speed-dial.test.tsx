import {
  act,
  createMockState,
  createMockStore,
  fireEvent,
  render,
  screen,
} from '@/tests/utils';
import * as Dialog from '../../feedback/dialog';
import {SpeedDialNav} from './speed-dial';

const dialogSpy = jest.spyOn(Dialog, 'Dialog');
const mockPush = jest.fn();

jest.mock('next/router', () => ({
  useRouter() {
    return {
      push: mockPush,
    };
  },
}));

beforeEach(() => {
  jest.clearAllMocks();
});

function hoverSpeedDial() {
  const speedDial = screen.getByRole('button', {name: 'speed-dial-menu'});
  fireEvent.mouseOver(speedDial);
}

function expectDialogToBeGone() {
  expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
}

describe('Speed dial', () => {
  it(`items render`, () => {
    render(<SpeedDialNav />);
    screen.getByRole('button');
    const items = screen.getAllByRole('menuitem');
    expect(items).toHaveLength(3);
  });
  it('has theme toggle', () => {
    const state = createMockState({app: {theme: 'dark'}});
    const store = createMockStore(state);
    const {unmount} = render(<SpeedDialNav />, store);
    let toggleButton = screen.getByRole('menuitem', {name: 'toggle-theme'});
    fireEvent.click(toggleButton);
    expect(store.getState().app.theme).toBe('light');
    unmount();
    render(<SpeedDialNav />, store);
    toggleButton = screen.getByRole('menuitem', {name: 'toggle-theme'});
    fireEvent.click(toggleButton);
    expect(store.getState().app.theme).toBe('dark');
  });
  it('has home navigation', () => {
    render(<SpeedDialNav />);
    hoverSpeedDial();
    const homeButton = screen.getByRole('menuitem', {name: 'home'});
    fireEvent.click(homeButton);
    expect(dialogSpy).toHaveBeenCalledTimes(1);
    expect(dialogSpy.mock.calls[0][0]).toMatchSnapshot(
      'home confirmation dialog'
    );
    act(() => {
      dialogSpy.mock.calls[0][0].onCancelCallback();
    });
    expectDialogToBeGone();
    fireEvent.click(homeButton);
    act(() => {
      dialogSpy.mock.calls[0][0].onConfirmCallback();
    });
    expectDialogToBeGone();
    expect(mockPush).toHaveBeenCalledWith('/');
  });
  it('can restart round', () => {
    const state = createMockState({
      game: {
        players: ['player2', 'player1'],
        rounds: {
          total: 3,
          current: 2,
          progress: 1,
        },
        scores: [
          [
            {name: 'player2', score: 0},
            {name: 'player1', score: 0},
          ],
          [{name: 'player2', score: 0}],
        ],
      },
    });
    const store = createMockStore(state);
    render(<SpeedDialNav />, store);
    hoverSpeedDial();
    const restartButton = screen.getByRole('menuitem', {name: 'restart'});
    fireEvent.click(restartButton);
    expect(dialogSpy).toHaveBeenCalledTimes(1);
    expect(dialogSpy.mock.calls[0][0]).toMatchSnapshot(
      'restart confirmation dialog'
    );
    act(() => {
      dialogSpy.mock.calls[0][0].onCancelCallback();
    });
    expectDialogToBeGone();
    fireEvent.click(restartButton);
    act(() => {
      dialogSpy.mock.calls[0][0].onConfirmCallback();
    });
    expectDialogToBeGone();
    expect(store.getState().game).toMatchSnapshot('reset round');
  });
});
