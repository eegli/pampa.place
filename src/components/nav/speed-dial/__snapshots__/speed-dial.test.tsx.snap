// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Speed dial can restart round: reset round 1`] = `
Object {
  "mapId": "test-my-map",
  "mapName": "my map",
  "players": Array [
    "player1",
    "player2",
  ],
  "rounds": Object {
    "current": 2,
    "progress": 0,
    "total": 3,
  },
  "scores": Array [
    Array [
      Object {
        "name": "player2",
        "score": 0,
      },
      Object {
        "name": "player1",
        "score": 0,
      },
    ],
    Array [],
  ],
  "status": "PENDING_PLAYER",
  "timeLimit": 60,
}
`;

exports[`Speed dial can restart round: restart confirmation dialog 1`] = `
Object {
  "confirmMessage": "Restart round",
  "infoMessage": "This will reset the current round progress for all players. The first player will start the current round again in a new location.",
  "onCancelCallback": [Function],
  "onConfirmCallback": [Function],
  "title": "Restart round?",
}
`;

exports[`Speed dial has home navigation: home confirmation dialog 1`] = `
Object {
  "confirmMessage": "Abort game",
  "onCancelCallback": [Function],
  "onConfirmCallback": [Function],
  "title": "Abort the game and return home?",
}
`;
